Generate Firefox Desktop front-end bindings with WebIDL

linacambridge opened the pull request:
This PR is a first cut at a UniFFI backend for Firefox! It's a work in progress, but can generate working bindings now, so I wanted to get it up for review. I'd love your comments! 💜

The backend works a little differently than the others. Firefox expects one header file per interface (and a separate one for the namespace that holds top-level functions), so UniFFI emits multiple `.h` and `.cpp` files. There's also complexity around handling different types. Some are reflected differently depending on whether they're arguments, type parameters, or local variables; some are returned directly and others are passed as out parameters; and the WebIDL implementation signatures have extra leading and trailing arguments. All this makes `macros.cpp` pretty gnarly 🙀

`SharedHeaderTemplate.h` contains the `RustBuffer` helpers and serialization logic. The ideas are the same as the Swift, Kotlin, and Python backends, but the implementation is a pile of C++ code. I'll add some more comments throughout it in the coming days, so that it's a bit easier to digest. This part can also be reused wholesale for our pure-C++ Nimbus interface, there's not much that's WebIDL specific here.

# Getting this building

The setup for this is a little janky for now, but here's how to get this running! You'll need a full (not artifact) build of Firefox, since this touches lots of C++ and Rust code.

1. `git checkout gecko` to switch to the Gecko branch.
2. Make a new folder for the bindings: `mkdir rondpoint-gecko`
3. Generate the Gecko bindings for Rondpoint: `cd uniffi_bindgen; and cargo run -- generate ../examples/rondpoint/src/rondpoint.idl -l gecko -o ../rondpoint-gecko`
4. Symlink the generated WebIDL file into `dom/chrome-webidl` in m-c: `ln -s /uniffi-rs/rondpoint-gecko/Rondpoint.webidl /gecko/dom/chrome-webidl/Rondpoint.webidl`
5. Add `Rondpoint.webidl` to `dom/chrome-webidl/moz.build`, in the `WEBIDL_FILES` section.
6. Symlink the `rondpoint-gecko` folder from step (2) into an existing (or new!) directory in m-c that has a `moz.build` file. Any one will do; I picked `toolkit/components/extensions/storage` for no other reason than it's a pretty small directory and makes it easy to do directory-specific builds.
7. Symlink `uniffi-rs` into m-c.
8. Add `uniffi-example-rondpoint` as a dependency to `toolkit/library/rust/shared/Cargo.toml`, as a `path` dependency inside the symlink from step (7).
9. Add `extern crate uniffi_rondpoint` to `toolkit/library/rust/shared/lib.rs`.
10. Run `mach vendor rust`.
11. Add the generated sources and xpcshell test manifest to the `moz.build` in the folder from step (6), like this:

```py
EXPORTS.mozilla.dom += [
    'rondpoint-gecko/Retourneur.h',
    'rondpoint-gecko/Rondpoint.h',
    'rondpoint-gecko/RondpointShared.h',
    'rondpoint-gecko/Stringifier.h',
]

UNIFIED_SOURCES += [
    'rondpoint-gecko/Retourneur.cpp',
    'rondpoint-gecko/Rondpoint.cpp',
    'rondpoint-gecko/Stringifier.cpp',
]

XPCSHELL_TESTS_MANIFESTS += [
    'uniffi-rs/examples/rondpoint/tests/gecko/xpcshell.ini'
]
```

And then run `mach build`, and you should be good to go! You can run `mach xpcshell-test toolkit/components/extensions/storage/uniffi-rs/examples/rondpoint/tests/gecko/xpcshell.ini` after the build is done! 😊

# Hacks

I had to make two changes to the `Cargo.toml`s in UniFFI to get this building in Firefox, which we'll want to fix before merging:

1. Remove the top-level `Cargo.toml`, so that I could symlink `uniffi-rs` into my m-c tree and add `uniffi-example-rondpoint` as a `path` dependency. I'm not sure this is necessary; if we can avoid symlinking `uniffi-rs` into m-c, we can do without.
2. Remove the `cargo_metadata` dependency. When I tried building Firefox with the version of `cargo_metadata` from Crates.io, I got a panic in one of its `derive` macros. Looks lik that was fixed in oli-obk/cargo_metadata#123 by bumping the Serde version, but that hasn't been published yet, so `mach vendor rust` doesn't know to pull the new version of Serde in. But changing `cargo_metadata` to a Git dependency confused Cargo: `mach vendor rust` didn't vendor _any_ of UniFFI's dependencies, and then `mach build` complained that it couldn't resolve the versions properly. I'm not sure what's going on there, and didn't dig too much more into it.

I also had to change the `crate-type` to `["cdylib", "lib"]`, so that we could link the crate into `gkrust`. If it's just `cdylib`, `extern crate` won't find it.

rfk commented on uniffi_bindgen/src/bindings/gecko/templates/WebIDLTemplate.webidl:
```diff
@@ -0,0 +1,67 @@
+  {#-
+  // We'll need to figure out how to handle async methods. One option is
+  // to declare them as `async foo()`, or an `[Async]` or `[BackgroundThread]`
+  // attribute in the UniFFI IDL. Kotlin, Swift, and Python can ignore that
```
Kotlin might like to automatically dispatch such methods via coroutine or similar, I think this could be worthwhile in more contexts than just Gecko.

linacambridge replied:
Ooooh, yes, great point! And maybe Swift can use a dispatch queue internally, and Python could use `asyncio`! 🚀

For the Nimbus pure-C++ API, we can punt on figuring out the exact details, and leave it up to a wrapper to initialize it and make network requests on a background task queue, just like we do in a-c and Swift now. But it’s definitely worth exploring this, and we’ll certainly want to have a story here before we transition the Nimbus JS SDK to Rust.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+/// Returns a dummy empty value for the given type. This is used to
+/// implement the `cpp::bail` macro to return early if a WebIDL function or
+/// method throws an exception.
+pub fn default_return_value_cpp(type_: &Type) -> Option<String> {
```
It’s unfortunate that this has to exist. Basically, if you want to throw an error from inside a WebIDL method with a signature like `int32_t Foo(ErrorResult& aRv)`, you can call `aRv.Throw(...)` in the method, but you still need to return _some_ value—even though it’ll get ignored by the bindings code. Methods that return results via out params don’t need to do this; they can just do `return;` and not set the param.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+
+/// A shared header file that's included by all our bindings. This defines
+/// common serialization logic and `extern` declarations for the FFI. Note that
+/// the bindings always include this header file, never the other way around.
```
...Because otherwise, we’d get a circular dependency when the bindings layer includes one of our interface headers in its header or implementation.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+#[template(syntax = "c", escape = "none", path = "NamespaceHeaderTemplate.h")]
+pub struct NamespaceHeader<'config, 'ci, 'functions> {
+    config: &'config Config,
+    ci: &'ci ComponentInterface,
```
It may be worth having these structs hold just the pieces they need from the `ComponentInterface`, like the namespace name, instead of the full interface _and_ the specific thing they’re rendering.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+            Type::Error(name) => panic!("[TODO: type_webidl({:?})]", type_),
+            Type::Optional(type_) => format!("{}?", type_webidl(type_)?),
+            Type::Sequence(type_) => format!("sequence<{}>", type_webidl(type_)?),
+            Type::Map(type_) => format!("record<DOMString, {}>", type_webidl(type_)?),
```
Terminology clarification that probably ought to go in a comment: what UniFFI calls a “record”, Firefox calls a “dictionary”. What UniFFI calls a “map”, Firefox calls a “record”. What UniFFI calls an “optional”, Firefox calls “nullable” (“optional” means something different in WebIDL/Firefox).

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+            FFIType::Float64 => "double".into(),
+            FFIType::RustBuffer => "RustBuffer".into(),
+            FFIType::RustString => "char*".into(),
+            FFIType::RustError => "NativeRustError".into(),
```
```suggestion
            FFIType::RustError => "RustError".into(),
```

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+        })
+    }
+
+    pub fn ret_type_ffi(type_: &Type) -> Result<String, askama::Error> {
```
Needs a comment—this is used in `macros.cpp`, to generate the specialization declaration for lifting return values from the FFI call, and for declaring the return type of an FFI call.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+            Type::Float32 => "float".into(),
+            Type::Float64 => "double".into(),
+            Type::Boolean => "bool".into(),
+            Type::String => "nsAString".into(),
```
This isn’t conceptually right (it should be `nsString`), but it _is_ the type we need to specialize `ViaFfi::Lift` and `Lower`. That might need to go into its own method...

And also, `type_cpp` needs a better name—something that captures this is the type expressed in a type parameter or when declaring a value.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/gen_gecko.rs:
```diff
@@ -0,0 +1,355 @@
+
+    pub fn enum_variant_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        // TODO: Make sure this does the right thing for hyphenated variants.
+        // Example: "bookmark-added" becomes `Bookmark_added`.
```
Let’s add a test case for hyphenated enum variants to UniFFI, and verify it gets reflected correctly in the binding layer.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/mod.rs:
```diff
@@ -0,0 +1,151 @@
+/// Generate uniffi component bindings for Gecko.
+///
+/// Bindings to a Rust interface for Gecko involves more than just generating a
+/// `.cpp` file.
```
Let’s move the comment about `Codegen.py` out of `generate_bindings` and up here.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/InterfaceHeaderTemplate.h:
```diff
@@ -0,0 +1,55 @@
+class {{ obj.name()|class_name_cpp }} final : public nsISupports, public nsWrapperCache {
+ public:
+  NS_DECL_CYCLE_COLLECTING_ISUPPORTS
+  NS_DECL_CYCLE_COLLECTION_SCRIPT_HOLDER_CLASS({{ obj.name()|class_name_cpp }})
```
Technically, since we only ever return new objects, we might not need cycle collection...but we do need to cycle-collect `mGlobal` if we hold onto it, and we may need it once we start returning references, _and_ also all this might go away and get replaced with XPIDL—so for now, copypasta it is! 🍝😁

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/InterfaceTemplate.cpp:
```diff
@@ -0,0 +1,97 @@
+  );
+  {%- if cons.throws().is_some() %}
+  if (err.mCode) {
+    aRv.ThrowOperationError(err.mMessage);
```
Error types are fun. Firefox has a hard-coded list of DOM exceptions (there aren’t that many! And I think the guidance for WebIDL APIs is to use a limited set of built-in exceptions, like `AbortError`, `InvalidStateError`, etc.) that can be thrown on an `ErrorResult&`...which is kind of at odds with how we expose errors. We might need to use `jsapi` here to make a custom error object, and then call `aRv.ThrowJSException`.

Propagating errors with all the right type information is still a big `TODO` for this PR.

linacambridge replied:
I filed https://github.com/mozilla/uniffi-rs/issues/295.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/InterfaceTemplate.cpp:
```diff
@@ -0,0 +1,97 @@
+  {% call cpp::decl_method_args(meth) %}
+) {
+  if (mHandle == -1) {
+    aRv.ThrowOperationError("Can't use destroyed handle");
```
Every method has a `[Throws]` annotation because it can throw one of these internal errors. Let’s drop this check (and setting the handle to `-1`) entirely. I think the `ConcurrentHandleMap` will fail any calls that try to use a destroyed handle, and this situation should never happen, anyway—when the destructor for a WebIDL object runs, it’s guaranteed that JS has GC’ed it.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+
+namespace mozilla {
+namespace dom {
+// namespace {{ ci.namespace() }} {
```
It would be nice if we could namespace the UniFFI bindings, but Firefox’s binding layer specifically expects all interface implementations to live in `mozilla::dom`.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+  /// on the reader.
+  template <typename T>
+  void ReadRawString(
+      T& aString,
```
Using an out parameter here because `AString` types (`nsACString` and `nsAString`) can’t be returned by value.

nnethercote replied:
This seems like a good comment to put in a comment :)

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(T);
+    AssertInBounds(newOffset);
+    mBuffer.SetLength(newOffset.value());
```
Add a comment: we’ve reserved the capacity in `mBuffer` to hold the entire object, so this won’t allocate, but we still need to update the length or we’ll get an out-of-bounds exception indexing into the array.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+template <typename T, typename FfiType>
+struct ViaFfi {
+  static bool Lift(const FfiType& aLowered, T& aLifted) = delete;
+  static FfiType Lower(const T& aLifted) = delete;
```
This returns `FfiType` instead of using an out param because we call `ViaFfi<T, U>::Lower(T)` for each argument in the argument list when calling an FFI function. An out param would mean we need to allocate some temporary stack variables to hold them, and I’ve nooo idea how to express all that in Askama 🤯 Fortunately, FFI types are all primitives, so there’s no fancy copy constructor or move shenanigans going on here, as would be the case with `Lift`.

nnethercote replied:
Another good thing to put in a comment.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+  }
+
+  static bool ReadFrom(Reader& aReader, nsACString& aValue) {
+    aValue.Truncate();
```
Just in case the out param has garbage in it.

nnethercote replied:
Again: put this in a comment :)

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+    RustError err = {0, nullptr};
+    char* lowered = {{ ci.ffi_string_alloc_from().name() }}(aLifted.BeginReading(), &err);
+    if (err.mCode) {
+      MOZ_ASSERT(false, "Failed to copy narrow string to Rust string");
```
`MOZ_ASSERT` doesn’t fire in release (`MOZ_RELEASE_ASSERT` does, though), which seems fine for things that aren’t security-sensitive—Rust will take care of crashing the app if it gets a `nullptr` for a string, or a weird serialization.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+};
+
+template <>
+struct ViaFfi<nsAString, char*> {
```
Technically, we can specialize `ViaFfi` for `nsString`, too—I didn’t because we only lift `char *`s into `nsAString&` out params, not `nsString`s on the heap or `nsAutoString`s on the stack.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+    RustError err = {0, nullptr};
+    char* lowered = {{ ci.ffi_string_alloc_from().name() }}(utf8.BeginReading(), &err);
+    if (err.mCode) {
+      MOZ_ASSERT(false, "Failed to copy wide string to Rust string");
```
```suggestion
      MOZ_ASSERT(false, "Failed to make Rust string from nsAString");
```

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+  static bool ReadFrom(Reader& aReader, dom::Nullable<T>& aValue) {
+    uint8_t hasValue = aReader.ReadUInt8();
+    if (hasValue != 0 && hasValue != 1) {
+      return false;
```
Assert about the unexpected tag here...

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+      MOZ_ASSERT(false);
+      return false;
+    }
+    {{ ci.ffi_bytebuffer_free().name() }}(aLowered);
```
Freeing a `const RustBuffer&` seems sketchy, even though it’s not actually `const`. Since we’re not inheriting from `Serializable`, just specializing, we can probably change this signature to `Lift(RustBuffer&)`, and the compiler will pick this overload instead of the deleted function in the base template.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,795 @@
+struct Serializable {
+  /// Returns the size of the serialized value, in bytes. This is used to
+  /// calculate the allocation size for the Rust byte buffer.
+  static size_t Size(const T& aValue) = delete;
```
I wonder if we should have this return a `CheckedInt<size_t>`, and only do the unwrapping and `MOZ_RELEASE_ASSERT(size.isValid())` once, when we actually allocate the buffer.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/WebIDLTemplate.webidl:
```diff
@@ -0,0 +1,67 @@
+  {%- endfor %}
+
+  {% for meth in obj.methods() -%}
+  [Throws]
```
See note above; get rid of the `mHandle != -1` check and then make this conditional on `method.throws()`, just like the others.

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/macros.cpp:
```diff
@@ -0,0 +1,120 @@
+  {%- endmatch -%}
+{%- endmacro -%}
+
+{# /* Declares a list of arguments for a WebIDL constructor. A constructor takes
```
These lists look quite hideous in the generated code, but I haven’t figured out the right mix of `%}` and `-%}` to make Askama strip the right amount of whitespace so that argument lists look pretty. The fixed arguments, like `GlobalObject`, and inserting the right number of commas make this tricky. They are readable as they are, so maybe that’s good enough?

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/macros.cpp:
```diff
@@ -0,0 +1,120 @@
+      {%- match ReturnPosition::for_type(type_) -%}
+        {%- when ReturnPosition::OutParam with (type_) -%},
+        {%- else -%}
+          {%- if !args.is_empty() %}, {% endif -%}
```
All this just to figure out if we need a trailing comma or not... 😭

linacambridge commented on uniffi_bindgen/src/bindings/gecko/templates/macros.cpp:
```diff
@@ -0,0 +1,120 @@
+  {% match ReturnPosition::for_type(type_) -%}
+  {%- when ReturnPosition::OutParam with (type_) -%}
+  DebugOnly<bool> ok_ = detail::ViaFfi<{{ type_|type_cpp }}, {{ type_|ret_type_ffi }}>::Lift({{ var }}, aRetVal_);
+  MOZ_ASSERT(ok_);
```
We should have already asserted before returning `false` in the serializers—maybe this deserves a `MOZ_RELEASE_ASSERT`?

nnethercote commented on examples/rondpoint/tests/gecko/test_rondpoint.js:
```diff
@@ -0,0 +1,141 @@
+add_task(async function test_rondpoint() {
```
What is `add_task`? I can see it used in this file, but I can't see it anywhere else.

mhammond replied:
I imagine the idea is to run these tests using `mach xpcshell-test` somehow - no idea what process lina has in mind for having mach understand them though

linacambridge replied:
Yep, this file is an xpcshell test!

> no idea what process lina has in mind for having mach find them though

Haha, my workflow so far has been “symlink `rondpoint` into m-c and add it to `moz.build` with all the other stuff”, but that’s not going to fly for anything more than a demo. Still, if you want to take it for a spin locally, there are only 11 easy steps! 😆

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+extern "C" {
+
+struct RustBuffer {
+  int64_t mLen;
```
@rfkelly was talking about changing buffer lengths to `int32_t`.

rfk replied:
In fact this chance has now landed on the `main` branch.

linacambridge replied:
Yep, the length is now `int32_t` and there's also a new capacity field.

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+  }
+
+  /// Helpers to read fixed-width primitive types at the current offset.
+  /// Fixed-width integers are read in big endian order.
```
I always find endianness tricky to think about, and I like to document things carefully. With that in mind, for me this comment raises as many questions as it answers. How/why are they read in big-endian order? Is big-endian used for all serialization in `uniffi-rs`? Some more explanation would be helpful.

(It might not hurt to mention that "network order" == "big-endian"... I had to google that to remind myself.)

rfk replied:
> Is big-endian used for all serialization in uniffi-rs?

It is, and it's probably worth a separate doc for the serialization format as a whole, since it's rather ad-hoc.

Some light background that might only be known to myself: I arbitrarily decided that everything should be big-endian when writing the first version of the serialization code, and I've no particular attachment to that decision. I wonder if it would be better to explicitly commit to using the native machine endianness, since we only ever expect to be serializing between two codebases running on the same machine (in the same process no less).

linacambridge replied:
I clarified in the functions themselves that this is because the serialization format uses big endian (I don't think the caller needs to know that one way or the other, any more than it needs to know we serialize floats and doubles by their bit pattern), and I also agree about documenting the format, maybe as part of the [ADR that introduced it](https://github.com/mozilla/uniffi-rs/blob/757a028278f8c8d81754e6a32ea3844012669deb/docs/adr/0002-serialize-complex-datatypes.md), or as an appendix in the tutorial! 👍🏻

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+  // This code uses `memcpy` and other eye-twitchy patterns because it
+  // originally wrote values directly into a `RustBuffer`, instead of
+  // an intermediate `nsTArray`. Once #251 is fixed, we can return to
+  // doing that, and remove `ToRustBuffer`.
```
The `Reader` code also uses `memcpy` but it doesn't have a comment like this. Should it?

Also, there should be a mention of big-endian encoding to match the one in `Reader`.

linacambridge replied:
I actually removed this comment because `Writer` writes directly into a `RustBuffer`, since it's now growable and we can avoid copying everything into the `nsTArray`, then copying everything again into a `RustBuffer`.

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+  /// this is just the length. For UTF-16 strings, which must be converted to
+  /// UTF-8, this can be up to three times the length. Once the closure tells us
+  /// how many bytes it's actually written, we can write the length prefix, and
+  /// advance the current offset.
```
Why does `ReadRawString` only support UTF-8 when this supports UTF-8 and UTF-16?

linacambridge replied:
It supports reading into a UTF-8 and a UTF-16 string, too—but the bytes in the `RustBuffer` are always UTF-8 encoded, to match Rust strings. But I realized the signatures for `ReadRawString` and `WriteRawString` were super confusing, especially with the lambda doing strange things, so I flattened them into `ReadCString` and `ReadString` methods. It's a little more code, but a lot more readable, so seems like worth the trade-off.

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+struct Serializable {
+  /// Returns the size of the serialized value, in bytes. This is used to
+  /// calculate the allocation size for the Rust byte buffer.
+  static CheckedInt<size_t> Size(const T& aValue) = delete;
```
Why is the return type `CheckedInt`?

linacambridge replied:
I was using `CheckedInt` throughout to avoid accidental overflows when serializing, but with the `RustBuffer` changes on `main` now, `Size` isn't necessary anymore. I think we can remove `CheckedInt` from the other uses, too, since `RustBuffer`s can only hold an `int32_t` worth of bytes.

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(float, ReadFloat, WriteFloat);
+UNIFFI_SPECIALIZE_SERIALIZABLE_PRIMITIVE(double, ReadDouble, WriteDouble);
+
+/// Booleans are passed as unsigned integers over the FFI, because JNA doesn't
```
Change "unsigned integers" to "Uint8"s... I was briefly confused because I erroneously assumed a UInt32.

linacambridge replied:
Ahh, oops, good call! Done! I also called out that it's an implementation detail of UniFFI, instead of mentioning JNA specifically.

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h:
```diff
@@ -0,0 +1,794 @@
+
+template <>
+struct ViaFfi<nsACString, char*> {
+  static MOZ_MUST_USE bool Lift(const char*& aLowered, nsACString& aLifted) {
```
`[[nodiscard]]` is now preferred to `MOZ_MUST_USE`. See [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1571631).

nnethercote commented on uniffi_bindgen/src/bindings/gecko/templates/WebIDLTemplate.webidl:
```diff
@@ -0,0 +1,69 @@
+  // We'll need to figure out how to handle async methods. One option is
+  // to declare them as `async foo()`, or an `[Async]` or `[BackgroundThread]`
+  // attribute in the UniFFI IDL. Kotlin, Swift, and Python can ignore that
+  // anno; Gecko will generate a method that returns a `Promise` instead, and
```
Is "anno" short for "annotation"? If so, I've not seen that abbreviation before. Probably best to write the full word.

nnethercote reviewed (commented):
This looks good to me. A lot of boilerplate-y stuff, but that's unavoidable, and it's easy to read and well-commented. I have a few nits, mostly about improving comments.

linacambridge commented:
OK, I think I've addressed all of @nnethercote's feedback, and rebased all the latest changes!

@rfk, would you mind giving this another quick glance, too? Thanks!

rfk reviewed (approved):
This looks really good.

I mean, I kinda glazed over a bit at the intricacies of the C++ code because I trust that more experienced folks are looked at it, but I'm confident I could use the guideposts here to dig in more to the details if needed!

I note that the generated bindings occupy a kind of global namespace, e.g. we get a class `mozilla::dom::Optionnuer` in a header file conditioned on `#ifndef mozilla_dom_Optionneur`. Are we able to add the uniffi package name as a level here, to give e.g. `mozilla::dom::Rondpointer::Optionnuer` and similar? I'm concerned about accidental name conflicts in future, but I'm also cautious about potential limitations of the way Firefox generates the bindings.

Similarly, if we had multiple uniffi libraries being build together in Firefox, would they run into trouble by both declaring `RustBuffer` structs etc? I think maybe it's OK because the `RondpointShared.h` file is only included by its `.cpp` files, not by any other header files. But I'm no expert, so kind of just looking for explicit re-assurance here :-)

A couple of other broad nits I noticed:

* When running `cargo test` in this branch, I get quite a few warnings about unused symbols which it's probably worth cleaning up.
* Running `uniffi-bindgen` via the `cargo run --` method you describe in the PR works, but if I try to build and run the executable via `./target/debug/uniffi-bindgen` using the same arguments, it produces an error `'Missing $CARGO_MANIFEST_DIR, cannot load config file: NotPresent'`. We should expect that some consumers will use an external `uniffi-bindgen` rather than building and running it from within this repo.

rfk commented on uniffi/src/testing.rs:
```diff
@@ -56,6 +56,8 @@ pub fn run_foreign_language_testcase(pkg_dir: &str, idl_file: &str, test_file: &
 /// Internally, this function does a bit of caching and concurrency management to avoid rebuilding
 /// the component for multiple testcases.
 pub fn ensure_compiled_cdylib(pkg_dir: &str) -> Result<String> {
+    unimplemented!()
```
I think this is causing CI test failures.

linacambridge commented:
> I mean, I kinda glazed over a bit at the intricacies of the C++ code because I trust that more experienced folks are looked at it, but I'm confident I could use the guideposts here to dig in more to the details if needed!

Thank you so much for digging into this, Ryan, that's mainly what I wanted from your review! 🚀

> I note that the generated bindings occupy a kind of global namespace, e.g. we get a class `mozilla::dom::Optionnuer` in a header file conditioned on `#ifndef mozilla_dom_Optionneur`. Are we able to add the uniffi package name as a level here, to give e.g. `mozilla::dom::Rondpointer::Optionnuer` and similar?

I don't think so 😞 I covered this in the Theory of Operation—which would be nice to convert into Markdown and check in to the repo, adding that to my list for this week ☑️ —but the binding declarations generated by Firefox's `Codegen.py` expect headers to be in `mozilla/dom/InterfaceName.h`, anda emit forward declarations that expect to find the interfaces and namespaces directly under `mozilla::dom`. (I think that also means type names will have to be globally unique, across UniFFI interfaces _and_ all the Web APIs currently in Gecko).

> Similarly, if we had multiple uniffi libraries being build together in Firefox, would they run into trouble by both declaring `RustBuffer` structs etc?

Yes, they would, because they're declared in the `extern "C"` block—thanks for noting this! I'll get that fixed up before landing, by prefixing `RustBuffer`, `ForeignBytes`, and `RustError` with the `ffi_namespace()`, like you do for the functions. It does mean we'll end up with duplication if we vendor multiple UniFFI components, but that's already the case today, and I suspect making cross-component helper deduping work is going to be more involved.

linacambridge commented on examples/arithmetic/Cargo.toml:
```diff
@@ -7,7 +7,7 @@ license = "MPL-2.0"
 
 [lib]
-crate-type = ["cdylib"]
+crate-type = ["cdylib", "lib"]
```
I'll go ahead and remove this since `arithmetic` doesn't have an xpcshell test, but we do need to do this in every crate we plan to vendor in m-c—since we're importing it as a Rust library via `extern crate`, not a `.dylib`.

linacambridge commented on uniffi/src/testing.rs:
```diff
@@ -10,7 +10,7 @@
 
 use anyhow::{bail, Result};
-use cargo_metadata::Message;
+// use cargo_metadata::Message;
```
Oops, I meant to revert these, thanks!

linacambridge replied:
@dmose took care of these, and the tests pass now! 🎉

linacambridge commented on uniffi_bindgen/src/bindings/gecko_js/gen_gecko_js.rs:
```diff
@@ -0,0 +1,463 @@
+/// determined by the `ComponentInterface`.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct Config {
+    // ...
```
Hmm, good question! There's limited customization that we can do, just because `Codegen.py` expects files and C++ namespaces to be laid out a certain way. I could imagine letting consumers specify a prefix, to make the WebIDL interface, namespace, dictionary, and type names globally unique, though—can you think of others we might want now?

linacambridge commented on uniffi_bindgen/src/bindings/gecko_js/gen_gecko_js.rs:
```diff
@@ -0,0 +1,463 @@
+    pub fn enum_variant_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        // TODO: Make sure this does the right thing for hyphenated variants.
+        // Example: "bookmark-added" should become `Bookmark_added`, because
+        // that's what Firefox's `Codegen.py` spits out.
```
I filed https://github.com/mozilla/uniffi-rs/issues/294 for this, but the consequences is that hyphenated names will produce a page-long compilation error because the enum variant that UniFFI generates doesn't exist in the variants that `Codegen.py` spits out 😬

linacambridge commented on uniffi_bindgen/src/bindings/gecko_js/templates/RustBufferHelper.h:
```diff
@@ -0,0 +1,630 @@
+/// `mozilla::Maybe<T>` under the hood, but their semantics are different.
+/// `Nullable<T>` means JS must pass some value for the argument or dictionary
+/// field: either `T` or `null`. `Optional<T>` means JS can omit the argument
+/// or member entirely.
```
It's a fun fact that's specific to Gecko, I'll either clarify or take this out entirely. In WebIDL, `Nullable<T>` means "a value of type T or `null`", and `Optional<T>` means "an argument or a dictionary member that can be omitted entirely". Something can be [both](https://searchfox.org/mozilla-central/rev/f27594d62e7f1d57626889255ce6a3071d67209f/dom/webidl/Animation.webidl#20-21) `Nullable` and `Optional`, in which case it gets reflected [like this](https://searchfox.org/mozilla-central/rev/f27594d62e7f1d57626889255ce6a3071d67209f/dom/animation/Animation.cpp#124-126).

linacambridge commented on uniffi_bindgen/src/bindings/gecko_js/templates/WebIDLTemplate.webidl:
```diff
@@ -0,0 +1,84 @@
+  // to declare them as `async foo()`, or an `[Async]` or `[BackgroundThread]`
+  // attribute in the UniFFI IDL. Kotlin, Swift, and Python can ignore that
+  // anno; Gecko will generate a method that returns a `Promise` instead, and
+  // dispatches the task to the background thread.
```
https://github.com/mozilla/uniffi-rs/issues/298, for posterity 😄

linacambridge commented on uniffi_bindgen/src/bindings/gecko_js/templates/macros.cpp:
```diff
@@ -0,0 +1,59 @@
+{# /* Calls an FFI function. */ #}
```
Seriously! It's a whole feast of curly braces, percents, angle brackets, and colons—lovely C++ 😭

rfk commented:
> prefixing RustBuffer, ForeignBytes, and RustError with the ffi_namespace(), like you do for the functions.
> It does mean we'll end up with duplication

I have perhaps too much faith in Sufficiently Smart Compilers, but hopefully these duplicated struct definitions would not result in duplicate code in the compiled artifacts, would they?

rfk commented:
> (I think that also means type names will have to be globally unique, across UniFFI interfaces and all the Web APIs currently in Gecko).

Hrm. I can see this being a problem (and worse, one that catches us by surprise). If you think there's scope for adding some sort of unique-ifying prefix to these names, it's probably worth putting the infra for that in place sooner rather than later.

linacambridge commented:
> I have perhaps too much faith in Sufficiently Smart Compilers, but hopefully these duplicated struct definitions would not result in duplicate code in the compiled artifacts, would they?

Good question—my knowledge of how C++ compilers inline templates, functions, and types is nonexistent 😭 Just like you suggested profiling the size impact of vendoring Nimbus and the bindings, it would be worth measuring the impact of vendoring multiple UniFFI components.

The one thing I'm worried about that might make inlining difficult is that the serialization helpers call into component-specific allocation functions. In `libxul` or an App Services-style megazord, `ffi_rondpoint_abc123_rustbuffer_alloc` does the exact same thing as `ffi_arithmetic_789xyz_rustbuffer_alloc`—but they're exported as two different symbols, and we end up generating two copies of the helpers wrapping them—`rondpoint_detail::{Reader, Writer, Serializable, ViaFfi}` and `arithmetic_detail::{Reader, Writer, Serializable, ViaFfi}`, plus all the specializations. Is Clang smart enough to recognize that, and optimize some (or all) those calls out? I don't know 😞

linacambridge commented:
> Hrm. I can see this being a problem (and worse, one that catches us by surprise). If you think there's scope for adding some sort of unique-ifying prefix to these names, it's probably worth putting the infra for that in place sooner rather than later.

Good idea. Let's do that now, so that this doesn't sneak up on us later.

linacambridge commented on examples/rondpoint/tests/gecko/test_rondpoint.js:
```diff
@@ -0,0 +1,141 @@
+  // Strings.
+  [
+    "",
+    "abc",
```
Done, and I went ahead and ported over the `Optionneur` tests while we were at it.

nnethercote reviewed (approved):
LGTM!

linacambridge commented:
Woohoo, thanks so much everyone! 😅 The latest push has the `definition_prefix` config option, prefixed names for the C structs (so we should be able to compile multiple UniFFIed components in Firefox), and a last-minute fix for nullable strings uncovered by the Rondpoint tests 👍🏻 (that example is paying for itself over and over!)

I'll go ahead and mash the "merge" button now :shipit: and we can file follow-ups for anything we forgot!
